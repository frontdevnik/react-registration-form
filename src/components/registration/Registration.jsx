import React from 'react';
import { useForm } from 'react-hook-form';
import { useSelector } from 'react-redux';

import {
  getFirstName,
  getEmail,
  getLastName,
  getPassword,
  getPhone,
  getRepeatPassword,
  getIsDiffPasswords,
} from '../../features/registration/selectors';

import MainInput from '../mainInput';

import {
  INVALID_NAME,
  INVALID_PASSWORD,
  INVALID_EMAIL,
  INVALID_PHONE,
} from '../../constants/validation-messages';
import {
  EMAIL_INPUT,
  NAME_INPUT,
  PASSWORD_INPUT,
  PHONE_INPUT,
} from '../../constants/validation-rules';

export default ({ onSubmit }) => {
  const defaultValues = {
    first_name: useSelector(getFirstName),
    last_name: useSelector(getLastName),
    email: useSelector(getEmail),
    phone: useSelector(getPhone),
    password: useSelector(getPassword),
    repeat_password: useSelector(getRepeatPassword),
  };
  const isDiffPasswords = useSelector(getIsDiffPasswords);

  const { register, handleSubmit, errors, reset } = useForm({
    mode: 'onBlur',
    defaultValues,
  });

  const handleReset = () => {
    reset(
      {},
      {
        errors: false,
      }
    );
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} className="form">
      <MainInput
        name="first_name"
        type="text"
        placeholder="First Name"
        required
        pattern={NAME_INPUT}
        errorMessage={INVALID_NAME}
        register={register}
        error={errors.first_name?.message}
      />
      <MainInput
        name="last_name"
        type="text"
        placeholder="Last Name"
        pattern={NAME_INPUT}
        errorMessage={INVALID_NAME}
        register={register}
        error={errors.last_name?.message}
      />
      <MainInput
        name="email"
        type="email"
        placeholder="Email"
        required
        pattern={EMAIL_INPUT}
        errorMessage={INVALID_EMAIL}
        register={register}
        error={errors.email?.message}
      />
      <MainInput
        name="phone"
        type="tel"
        placeholder="Phone Number"
        required
        pattern={PHONE_INPUT}
        errorMessage={INVALID_PHONE}
        register={register}
        error={errors.phone?.message}
      />
      <MainInput
        name="password"
        type="password"
        placeholder="Password"
        required
        pattern={PASSWORD_INPUT}
        errorMessage={INVALID_PASSWORD}
        register={register}
        error={errors.password?.message}
      />
      <MainInput
        name="repeat_password"
        type="password"
        placeholder="Repeat password"
        required
        pattern={PASSWORD_INPUT}
        errorMessage={INVALID_PASSWORD}
        register={register}
        error={errors.repeat_password?.message}
      />
      {isDiffPasswords && (
        <p className="datepicker-clear">Passwords don't match</p>
      )}
      <button className="waves-effect waves-light btn">submit</button>
      <input
        type="button"
        className="waves-effect waves-light btn secondary-content blue-grey"
        onClick={handleReset}
        value="reset"
      />
    </form>
  );
};
