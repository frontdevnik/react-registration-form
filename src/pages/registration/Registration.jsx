import React from 'react';
import { useDispatch } from 'react-redux';

import MainWrapper from '../../components/mainWrapper';
import Breadcrumbs from '../../components/navigations';
import Registration from '../../components/registration';

import {
  submitRegistrationForm,
  differentPasswords,
} from '../../features/registration/actions';

export default () => {
  const dispatch = useDispatch();

  const handleSubmit = (data) => {
    const { password, repeat_password } = data;

    password !== repeat_password
      ? dispatch(differentPasswords())
      : dispatch(submitRegistrationForm(data));
  };

  return (
    <MainWrapper>
      <Registration onSubmit={handleSubmit} />
      <Breadcrumbs />
    </MainWrapper>
  );
};
