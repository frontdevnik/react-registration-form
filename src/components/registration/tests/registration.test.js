import React from 'react';

import Registration from '../Registration';

import {
  stubRegistrationProps,
  registrationLinks,
} from './stubProps';

jest.mock('react-redux', () => ({
  useSelector: jest.fn((f) => f()),
}));

jest.mock('react-hook-form', () => ({
  useForm: () => ({
    register: jest.fn(),
    handleSubmit: jest.fn((f) => f()),
    reset: jest.fn(),
    errors: {},
  }),
}));

jest.mock('../../../features/registration/selectors');

const setUp = (props) => mount(<Registration {...props} />);

describe('<Registration />', () => {
  let component;
  let form;

  beforeEach(() => {
    component = setUp(stubRegistrationProps);
    form = component.find('form');
  });

  it('should render correctly', () => {
    expect(component.html()).toMatchSnapshot();
  });

  it('should form has correct class', () => {
    expect(form.hasClass('form')).toBeTruthy();
  });

  it('should call handleSubmit on submit form', () => {
    form.simulate('submit');
    expect(stubRegistrationProps.onSubmit).toHaveBeenCalled();
  });

  it('should all inputs render correctly', () => {
    component.find('input').forEach((node, i) => {
      if (!registrationLinks[i]) {
        return;
      }

      const nodeProps = node.props();
      const { name, type, placeholder } = registrationLinks[i];

      expect(nodeProps.name).toBe(name);
      expect(nodeProps.type).toBe(type);
      expect(nodeProps.id).toBe(name);
      expect(nodeProps.placeholder).toBe(placeholder);
      node.simulate('change');
      expect(stubRegistrationProps.onSubmit).toHaveBeenCalled();
    });
  });

  it('should call submitForm on button click', () => {
    const button = component.find('button');
    button.simulate('click');
    expect(stubRegistrationProps.onSubmit).toHaveBeenCalled();
  });

  it('should render reset button', () => {
    const resetInput = component.find('input').last();
    expect(resetInput.props().value).toBe('reset');
  });
});
