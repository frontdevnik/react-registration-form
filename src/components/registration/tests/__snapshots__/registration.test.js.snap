// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> should render correctly 1`] = `"<form class=\\"form\\"><div class=\\"input-field col s12\\"><input name=\\"first_name\\" type=\\"text\\" id=\\"first_name\\" placeholder=\\"First Name\\" class=\\"\\"><label for=\\"first_name\\"></label></div><div class=\\"input-field col s12\\"><input name=\\"last_name\\" type=\\"text\\" id=\\"last_name\\" placeholder=\\"Last Name\\" class=\\"\\"><label for=\\"last_name\\"></label></div><div class=\\"input-field col s12\\"><input name=\\"email\\" type=\\"email\\" id=\\"email\\" placeholder=\\"Email\\" class=\\"\\"><label for=\\"email\\"></label></div><div class=\\"input-field col s12\\"><input name=\\"phone\\" type=\\"tel\\" id=\\"phone\\" placeholder=\\"Phone Number\\" class=\\"\\"><label for=\\"phone\\"></label></div><div class=\\"input-field col s12\\"><input name=\\"password\\" type=\\"password\\" id=\\"password\\" placeholder=\\"Password\\" class=\\"\\"><label for=\\"password\\"></label></div><div class=\\"input-field col s12\\"><input name=\\"repeat_password\\" type=\\"password\\" id=\\"repeat_password\\" placeholder=\\"Repeat password\\" class=\\"\\"><label for=\\"repeat_password\\"></label></div><button class=\\"waves-effect waves-light btn\\">submit</button><input type=\\"button\\" class=\\"waves-effect waves-light btn secondary-content blue-grey\\" value=\\"reset\\"></form>"`;
